DISPLAY: match on defined names

andreasabel opened the pull request:
Re #2004: DISPLAY pragmas can now match on defined names.
This PR modifies the scope checker so as to accept defined names (functions, postulate, data & record types, primitives) as matchable (`DefP`).
Technically, we pass around a flag `displayLhs : Bool` that allows us to know whether we are scope checking the lhs of a DISPLAY pragma.
Rather than putting this into the environment (`TCEnv`), I am defining specific wrappers `CPattern` and `CLHSCore` that carry that flag.  (This seems more structured than just a global environment hosting a myriad of parameters of which always only a few apply to the current situation.)

Closes #7533 (and parts of #2004).

TODO (later):
- figure out whether we can also match on projections
- more tests

andreasabel committed:
Testsuite: restore name in succeed.agda-lib

andreasabel committed:
Re #2004: DISPLAY pragmas can now match on defined names

Closes #7533.

Squashed into this commit:

- Re #2004: IsBool DisplayLHS instead of Boolean

- Re #2004: get rid of ApplicativeDo

UlfNorell commented on src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs:
```diff
@@ -3086,35 +3113,59 @@ mergeEqualPs = go (empty, [])
+        ps2  <- toAbstract $ (fmap . fmap . fmap) (CPattern displayLhs) ps2
+        A.LHSProj (AmbQ ds) core <$> mergeEqualPs ps2
+
+    C.LHSWith core wps ps -> do  -- N.B.: This is an ApplicativeDo.
```
Why is it important that this is an ApplicativeDo? I would be a little concerned if the semantics relies on this to be the case!

andreasabel replied:
I can take this comment away, if it is irritating.  
This was a nested `liftA2`, but rather unreadable.  
`ApplicativeDo` shouldn't change the semantics, don't even know whether it makes things more performant.

UlfNorell replied:
I would get rid of the comment and the options pragma at the top. We don't care about restricting to the applicative interface anywhere else, nor do we have any evidence that it would be beneficial.

andreasabel replied:
Done.

UlfNorell reviewed (approved):
It might be worth adding a separate type for AreWeCheckingADisplayForm since the booleans show up in a lot of places.

andreasabel committed:
Agda.Utils.Monad: generalize when, unless; reexport Control.Monad

Since mtl-2.2 also reexports `Control.Monad`, we need to import the
`Control.Monad.*` modules qualified to avoid clashes with our own `when`
and `unless`.

andreasabel commented:
> It might be worth adding a separate type for AreWeCheckingADisplayForm since the booleans show up in a lot of places.

Ok, I introduced `IsBool DisplayLHS`.
I got a bit carried away generalizing `unless` from `Boolean`  to `IsBool b`,  so there is a bit commit reorganizing imports so that it does not conflict with `Control.Monad.unless`.

andreasabel commented:
A side effect of not distinguishing universes from definitions is that one can have DISPLAY forms for them:
```agda
postulate
  Type : Set₁

{-# DISPLAY Set = Type #-}

test : Type
test = Set
```
> Set₁ !=< Type when checking that the expression Set has type Type

But this is probably not worrisome.  DISPLAY is anyway unsound, so all bets are off.
